presto-main/src/main/java/io/prestosql/type/Timestamps.java

```diff
@@ -0,0 +1,342 @@
+        }
+
+        return POWERS_OF_TEN[toPrecision - fromPrecision];
+    }
+
+    public static long rescale(long value, int fromPrecision, int toPrecision)
+    {
+        if (value < 0) {
+            throw new IllegalArgumentException("value must be >= 0");
+        }
+
+        if (fromPrecision <= toPrecision) {
+            value *= scaleFactor(fromPrecision, toPrecision);
+        }
+        else {
+            value = roundDiv(value, scaleFactor(toPrecision, fromPrecision));
+        }
+
+        return value;
+    }
+
+    public static boolean timestampHasTimeZone(String value)
+    {
+        Matcher matcher = DATETIME_PATTERN.matcher(value);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException(format("Invalid timestamp '%s'", value));
+        }
+
+        return matcher.group("timezone") != null;
+    }
+
+    public static int extractTimestampPrecision(String value)
+    {
+        Matcher matcher = DATETIME_PATTERN.matcher(value);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException(format("Invalid timestamp '%s'", value));
+        }
+
+        String fraction = matcher.group("fraction");
+        if (fraction == null) {
+            return 0;
+        }
+
+        return fraction.length();
+    }
+
+    public static LocalDateTime toLocalDateTime(TimestampType type, ConnectorSession session, Block block, int position)
+    {
+        int precision = type.getPrecision();
+
+        long epochMicros;
+        int picosOfMicro = 0;
+        if (precision <= 3) {
+            epochMicros = scaleEpochMillisToMicros(type.getLong(block, position));
+        }
+        else if (precision <= MAX_SHORT_PRECISION) {
+            epochMicros = type.getLong(block, position);
+        }
+        else {
+            LongTimestamp timestamp = (LongTimestamp) type.getObject(block, position);
+            epochMicros = timestamp.getEpochMicros();
+            picosOfMicro = timestamp.getPicosOfMicro();
+        }
+
+        long epochSecond = scaleEpochMicrosToSeconds(epochMicros);
+        int nanoFraction = getMicrosOfSecond(epochMicros) * NANOSECONDS_PER_MICROSECOND + (int) (roundToNearest(picosOfMicro, PICOSECONDS_PER_NANOSECOND) / PICOSECONDS_PER_NANOSECOND);
+
+        Instant instant = Instant.ofEpochSecond(epochSecond, nanoFraction);
+        if (session.isLegacyTimestamp()) {
+            return LocalDateTime.ofInstant(instant, session.getTimeZoneKey().getZoneId());
+        }
+
+        return LocalDateTime.ofInstant(instant, ZoneOffset.UTC);
+    }
+
+    public static String formatTimestamp(int precision, long epochMicros, int picosOfMicro, ZoneId zoneId)
+    {
+        return formatTimestamp(precision, epochMicros, picosOfMicro, zoneId, TIMESTAMP_FORMATTER);
+    }
+
+    public static String formatTimestamp(int precision, long epochMicros, int picosOfMicro, ZoneId zoneId, DateTimeFormatter yearToSecondFormatter)
```

Comment:
nit: verify picosOfMicro is in appropriate range (primarily to catch negative values)